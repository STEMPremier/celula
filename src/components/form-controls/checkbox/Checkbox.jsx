/* eslint-disable no-console */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import './checkbox.less';

class Checkbox extends Component {
  state = {
    checked: false,
    showFocus: false,
  };

  componentDidMount() {
    const { checked } = this.props;

    if (checked) {
      this.setState({
        checked: true,
      });
    }
  }

  handleChange = event => {
    const { handleChange } = this.props;

    this.setState({
      checked: event.target.checked,
    });

    handleChange(event.target.value);
  };

  onMouseEnter = () => {
    this.setState(() => ({
      showFocus: true,
    }));
  };

  onMouseLeave = () => {
    this.setState(() => ({ showFocus: false }));
  };

  render() {
    const {
      className,
      disabled,
      errorMsg,
      formId,
      name,
      label,
      value,
    } = this.props;
    const { checked, showFocus } = this.state;
    const id = `${name}_${value}`;
    const classes = cx(
      'ce-checkbox',
      {
        'ce-checkbox--disabled': disabled,
        'ce-checkbox--individual-checkbox-error': errorMsg,
      },
      className,
    );

    return (
      <div
        className={classes}
        onMouseEnter={this.onMouseEnter}
        onMouseLeave={this.onMouseLeave}
      >
        <div className="ce-checkbox--box">
          <input
            checked={checked}
            disabled={disabled}
            form={formId}
            id={id}
            name={name}
            onChange={this.handleChange}
            type="checkbox"
            value={value}
          />
          <label htmlFor={id}>{label}</label>
          <div className="ce-checkbox--background" />
          {showFocus && <div className="ce-checkbox--focus" />}
          {errorMsg && (
            <div className="ce-checkbox--error-box-wrapper">
              <div className="ce-checkbox--arrow" />
              <div className="ce-checkbox--error-box">
                <div className="ce-checkbox--error-box-text">{errorMsg}</div>
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }
}

Checkbox.propTypes = {
  /**
   * Select the `<Checkbox />`.
   */
  checked: PropTypes.bool,
  /**
   * A class name added to the `<CheckboxGroup />`.
   */
  className: PropTypes.string,
  /**
   * Make the `<Checkbox />` inactive.
   */
  disabled: PropTypes.bool,
  /**
   * An error message to display for a `<Checkbox />'.
   */
  errorMsg: PropTypes.string,
  /**
   * The id of the form the `<Checkbox />` belongs to. This value is overwritten by any value passed from an enclosing `<CheckboxGroup />`.
   */
  formId: PropTypes.string,
  /**
   * A function to trigger when the state of the `<Checkbox />` changes.
   */
  handleChange: PropTypes.func,
  /**
   * This component is to be treated as a checkbox in `<FormControlGroup />`.
   * @ignore
   */
  isA: PropTypes.string, // eslint-disable-line react/no-unused-prop-types
  /**
   * The `<Checkbox />` label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Need to assign a matching name to each `<Checkbox />` if it is being used outside of the `<CheckboxGroup />`.  In  a `<CheckboxGroup />`, each individual `<Checkbox />` will have the same name, but that will come from the `<CheckboxGroup />`.
   */
  name: PropTypes.string,
  /**
   * The value of the `<Checkbox />`.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ]).isRequired,
};

Checkbox.defaultProps = {
  checked: false,
  className: '',
  disabled: false,
  errorMsg: '',
  formId: '',
  handleChange: () => {},
  isA: 'checkbox',
  name: '',
};

export default Checkbox;
