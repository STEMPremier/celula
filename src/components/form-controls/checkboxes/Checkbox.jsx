/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/no-unused-state */
/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import './checkboxes.less';

class Checkbox extends Component {
  state = {
    checked:
      this.props.defaultGroupValue &&
      !!this.props.defaultGroupValue.includes(this.props.value)
        ? this.props.defaultGroupValue
        : this.props.checkedIndividualDefault,
    // ? this.props.defaultGroupValue || this.props.checkedIndividualDefault
    // : '',

    // this.props.defaultGroupValue &&
    // !!this.props.defaultGroupValue.includes(this.props.value)
    //   ? this.props.defaultGroupValue
    //     this.props.checkedIndividualDefault,
  };

  handleChange = event => {
    this.setState({ checked: event.target.checked });
    // eslint-disable-next-line react/destructuring-assignment
    this.props.handleChange();
  };

  render() {
    const {
      className,
      disabled,
      error,
      name,
      label,
      value,
      form,
      checkedIndividualDefault,
    } = this.props;

    const { checked } = this.state;
    const id = `${name}_${value}`;
    const classes = cx(
      'ce-checkbox',
      {
        'ce-checkbox--disabled': disabled,
        'ce-checkbox--individual-checkbox-error': error,
      },
      className,
    );

    return (
      <div className={classes}>
        <div className="ce-checkbox--box">
          <input
            type="checkbox"
            value={value}
            id={id}
            checkedIndividualDefault={checkedIndividualDefault}
            checked={checked}
            onClick={this.handleChange}
            disabled={disabled}
            name={name}
            form={form}
          />
          <label htmlFor={id}>{label}</label>
          <div className="ce-checkbox--background" />
          {error && (
            <div className="ce-checkbox--error-box-wrapper">
              <div className="ce-checkbox--arrow" />
              <div className="ce-checkbox--error-box">
                <div className="ce-checkbox--error-box-text">{error}</div>
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }
}

Checkbox.propTypes = {
  /**
   * A class name added to the `<CheckboxGroup />`.
   */
  className: PropTypes.string,
  /**
   * The current checked state.
   */
  checkedIndividualDefault: PropTypes.bool,
  /**
   * Make the RadioGroup inactive.
   */
  disabled: PropTypes.bool,
  /**
   * Error message on single `<Checkbox />'
   */
  error: PropTypes.string,
  /**
   * Need to assign a matching name to each `<Checkbox />` if it is being used in the `<CheckboxGroup />`.  In  a `<CheckboxGroup />`, each individual `<Checkbox />` should have the same name.
   */
  name: PropTypes.string,
  /**
   * The handleChange function sets the checked state of the individual `<Checkbox />` and also trigger the handleChangeGroup of the `<CheckboxGroup />`.
   */
  handleChange: PropTypes.func,
  /**
   * The form value is over-written by any form value passed into the `<CheckboxGroup />`.
   */
  form: PropTypes.string,
  /**
   * The label you assign is the text that shows up next to each individual `<Checkbox />` button.
   */
  label: PropTypes.string.isRequired,
  /**
   * The value is not visible to the user, but rather it is the unique value passed when selecting each individual `<Checkbox />` button.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ]).isRequired,
};

Checkbox.defaultProps = {
  className: '',
  form: '',
  checkedIndividualDefault: false,
  name: '',
  disabled: false,
  handleChange: () => {},
  error: '',
};

export default Checkbox;
