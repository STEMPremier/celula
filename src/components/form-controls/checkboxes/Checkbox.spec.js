/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Checkbox from './Checkbox';

describe('<Checkbox />', () => {
  describe('render()', () => {
    test('render the CheckboxGroup component', () => {
      const wrapper = shallow(
        <Checkbox
          label="Checkbox Test Label 1"
          value="Checkbox Value Here 1"
        />,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
