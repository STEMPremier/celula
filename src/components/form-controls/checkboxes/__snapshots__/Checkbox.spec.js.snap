// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> render() render the CheckboxGroup 1`] = `
<div
  className="ce-checkbox"
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="ce-checkbox--box"
    >
      <input
        checked={false}
        disabled={false}
        form=""
        id="_Checkbox Value Here 1"
        name=""
        onChange={[Function]}
        type="checkbox"
        value="Checkbox Value Here 1"
      />
      <label
        htmlFor="_Checkbox Value Here 1"
      >
        Checkbox Test Label 1
      </label>
      <div
        className="ce-checkbox--background"
      />
    </div>
  </div>
</div>
`;

exports[`<Checkbox /> render() render the CheckboxGroup component 1`] = `
<div
  className="ce-checkbox"
>
  <div
    className="ce-checkbox--box"
  >
    <input
      checked={false}
      disabled={false}
      form=""
      id="_Checkbox Value Here 1"
      name=""
      onChange={[Function]}
      type="checkbox"
      value="Checkbox Value Here 1"
    />
    <label
      htmlFor="_Checkbox Value Here 1"
    >
      Checkbox Test Label 1
    </label>
    <div
      className="ce-checkbox--background"
    />
  </div>
</div>
`;
